client/my-sites/sharing/connections/service.jsx

```diff
@@ -132,65 +186,144 @@ module.exports = React.createClass( {
-		this.props.connections.once( 'create:error', this.onConnectionError );
-		this.props.onAddConnection( this.props.service );
-	},
-
-	disconnect: function( connections ) {
-		if ( 'undefined' === typeof connections ) {
-			// If connections is undefined, assume that all connections for
-			// this service are to be removed.
-			connections = serviceConnections.getRemovableConnections( this.props.service.ID );
-		}
-
-		this.setState( { isDisconnecting: true } );
-		this.props.connections.once( 'destroy:success', this.onDisconnectionSuccess );
-		this.props.connections.once( 'destroy:error', this.onDisconnectionError );
-		this.props.onRemoveConnection( connections );
-	},
-
-	refresh: function( connection ) {
+	refresh: function( oldConnection ) {
 		this.setState( { isRefreshing: true } );
 		this.props.connections.once( 'refresh:success', this.onRefreshSuccess );
 		this.props.connections.once( 'refresh:error', this.onRefreshError );
 
-		if ( ! connection ) {
+		if ( ! oldConnection ) {
 			// When triggering a refresh from the primary action button, find
 			// the first broken connection owned by the current user.
-			connection = serviceConnections.getRefreshableConnections( this.props.service.ID )[ 0 ];
+			oldConnection = this.getConnections().filter( ( connection ) => ( 'broken' === connection.status ) );
 		}
-		this.props.onRefreshConnection( connection );
+		this.refreshConnection( oldConnection );
 	},
 
 	performAction: function() {
-		const connectionStatus = serviceConnections.getConnectionStatus( this.props.service.ID );
+		const connectionStatus = this.getConnectionStatus( this.props.service.ID ),
+			removableConnections = this.getRemovableConnections( this.props.service.ID );
 
 		// Depending on current status, perform an action when user clicks the
 		// service action button
-		if ( 'connected' === connectionStatus && serviceConnections.getRemovableConnections( this.props.service.ID ).length ) {
-			this.disconnect();
-			analytics.ga.recordEvent( 'Sharing', 'Clicked Disconnect Button', this.props.service.ID );
+		if ( 'connected' === connectionStatus && removableConnections.length ) {
+			this.removeConnection( removableConnections );
+			this.props.recordGoogleEvent( 'Sharing', 'Clicked Disconnect Button', this.props.service.ID );
 		} else if ( 'reconnect' === connectionStatus ) {
 			this.refresh();
-			analytics.ga.recordEvent( 'Sharing', 'Clicked Reconnect Button', this.props.service.ID );
+			this.props.recordGoogleEvent( 'Sharing', 'Clicked Reconnect Button', this.props.service.ID );
 		} else {
-			this.connect();
-			analytics.ga.recordEvent( 'Sharing', 'Clicked Connect Button', this.props.service.ID );
+			this.addConnection();
+			this.props.recordGoogleEvent( 'Sharing', 'Clicked Connect Button', this.props.service.ID );
 		}
 	},
 
+	addConnection: function( service = this.props.service, keyringConnectionId, externalUserId = false ) {
+		const _this = this,
```

Comment:
This should be unnecessary since the `popupMonitor` event handler uses an arrow function (`this` still refers to the component instance).